### DISCUSSION TITLE
property initialization, and `apply` method call and change events

### DISCUSSION THREAD

**Discussion by derrell** [qooxdoo/qooxdoo#10449] [Ideas]
# Introduction
As I work on a completely reimplemented class/property system for qooxdoo, which eliminates the very difficult-to-maintain machine-generated code for each property, I am finding some details that are not well documented, leaving only the unit tests as the basis for ascertaining the intended functionality. In some cases, I'm unable to ascertain the actual intended actions. These tend to be edge cases. One of these cases pertains to a property's init value, and its `apply` method and change event. Specifically:

- what is a property's initial value, depending on its other attributes
- whether or not... and when... does the property's `apply` method get called
- whether the property's change event get fired
- for the `apply` method and the change event, what arguments are provided for `value` and `old`

Since the behavior is entirely undocumented, this discussion is to decide on the set of rules I will use to implement this in the new class/property system. These decisions may create some backward-compatibility changes that will then be documented in the qooxdoo v8 release notes.

# Traditional options for initialization and reinitialization
These are the "rules" for initialization that I have intuited from reading the documentation and code, and running the unit tests.

## Initial value of a property
1. A property may be assigned an initial value with the `init` key in the property configuration. The developer must be careful not to assign a reference value such as `new qx.ui.core.Widget()`, that is supposed to be unique to each instance, as the init value. The `new qx.ui.core.Widget()` code is executed only once, and the same reference will be assigned to each instance of the class defining the property. The `deferredInit` key was intended to assist in initializing a property with a reference value, but did not help in many cases.
2. If a property has in its configuration, `nullable : true`, but no `init` value, then the initial value will be `null`
3. Otherwise, if a property has in its configuration, `check : "Boolean"`, but no `init` value, then the initial value will be `false`.
4. If none of the above apply, then the initial value is `undefined`... but the getter may sometimes change that to `null` in some obscure and incomprehensible way. Also, see the discussion below, regarding `themeable` and `inheritable` property values.
5. A property of name `x` can be reinitialized by calling `initX()` This function may be passed a value, in which case, that value is used for the initialization. If no value is provided, then the above-described rules are used to set the property's value.

## Themeable and Inheritable property values
A property can be marked as `themeable : true` to indicate that its value may be provided by an appearance theme. Alternatively, a property can be marked as `inheritable : true` to indicate that its value may be taken from the property of the same name, in the layout parent of the property's instance, i.e., from the container in which the widget is placed.

When property `x`'s `resetX()` method is called, the property is reinitialized as follows:

1. If a theme value for this property has been provided by the theme, the theme's value is used
2. Otherwise, if an inherited value has been provided by the layout parent, the inherited value is used
3. In any other case, the init value, ascertained as described above, is used

## Setting a property's value
The value of a property can be changed by calling its setter, for example, `setX(value)`. This is referred to, internally, as setting the "user" value.

## Refreshing a property's value
The property's `refreshX()` method uses the above calculations to set a potentially new value for the property, as follows:

1. If the property has been assigned a "user" value with `setX(value)`, then that value is (re-)assigned to the property.
2. Otherwise, if there's a not-undefined initial value, as described above, that value is (re-)assigned to the property.
3. Otherwise, if the property is inheritable and there is a layout parent which has that same property name, its value is (re-)assigned to the property.
4. Otherwise, no change is made to the property value

## New feature in qooxdoo v8 class/property system: `initFunction`
As mentioned above, there is a problem using the `init` key to assign a reference value to a property, as the same reference will be used for all object instances. To solve that problem, a new property configuration key is added: `initFunction`. The value of this key is expected to be a function which takes no arguments, and returns the initial value to be assigned to the property of an object instance. This function is called immediately after a new object instance is created, and immediately before the object's constructor is called. This function is called and its return value also used, in preference to an `init` value, in all of the reinitialization processes described above.

# Calling a property's `apply` method and firing its event
This is where the rules become murky. Let's start with what seems consistent.

- The conditions for when a property's `apply` method (if it has one) is called, and when its change event is fired, are the same. There does not seem to be any time that an `apply` method is called but the change event is not fired; or vice versa.
- If a property has previously been given a value, and is then given a new value that is the same as the previously given value, the `apply` method is not called and the change event is not fired.
- If a property has previously been given a value, and is then given a new value that is different than the previously given value, the `apply` method is called and the change event is fired. The arguments to the `apply` method, and the data passed in the fired event, include `value`, the new value, and `old`, the prior value.

Where inconsistencies and ambiguities live, is what happens with the `apply` method and change event, in conjunction with `init` values, the `initX()` method, and the other reinitialization processes described above. There is an attempt, it seems, to *force call* the `apply` method and fire the change event upon the first time that a "user" value is applied to the property, with `setX(value)`, even if that provided `value` is the same as the property's `init` value. The unit tests, however, give us only ambiguous answers to both, when `apply` is called and the change event fired, but also, to what `old` value is provided.

# New rules
For the qooxdoo v8 implementation of classes/properties, I intend to implement to these rules dictating when the `apply` method is called and the change event generated, and what arguments are provided to them.

Firstly, the rules that seemed consistent previously, will remain exactly as they were. To wit:

- The conditions for when a property's `apply` method (if it has one) is called, and when its change event is fired, are the same. There does not seem to be any time that an `apply` method is called but the change event is not fired; or vice versa.
- If a property has previously been given a value, and is then given a new value that is the same as the previously given value, the `apply` method is not called and the change event is not fired.
- If a property has previously been given a value, and is then given a new value that is different than the previously given value, the `apply` method is called and the change event is fired. The arguments to the `apply` method, and the data passed in the fired event, include `value`, the new value, and `old`, the prior value.

My current plan is that the following additional rules will be applied, once the indicated questions are answered. These may not be entirely backward compatible with the qooxdoo v7 and older implementation.

- `init` values, and the return value from `initFunction`, are not validity checked via any provided`check` mechanism, nor passed through `transform` or other property functions that might throw an error. It is up to the developer to provide a legal value of the property as its `init` value or `initFunction` return value.
- `init` values and the return value from `initFunction` never cause the `apply` method to be called, nor the change event to be fired. They simply set the initial value of the property.
- The *force call* described above, upon first "user" `setX(value)` call, is **not** done when the old and new values are the same.

For initial values assigned via `initFunction`'s return value, I'd really like for the `apply` method to be called, or to allow it to be called at user preference, at one of the following times:

- immediately before the constructor is called
- immediately after the constructor is called
- via some new method call that the developer can call from within the constructor, e.g., `this.applyIinitValues()`. Since we're talking about initial values assigned via `initFunction`, the specified `initFunction` could even call a method, e.g., `this.requestApplyInit(propertyName)` requesting that its `apply` method be called when the constructor calls `this.applyInitValues()`

The problems are:

-  if called before the constructor, the `apply` method may not be ready to be called as it may require some initialization that would occur in the constructor... and there wouldn't yet be any listeners for the change event since the constructor hasn't even been called yet
- if called after the constructor is called, the constructor may have set new values for those properties, and the init values may no longer be valid. This could be overcome by noting whether a "user" value has been set, and not calling the `apply` method in that case.

That leaves the third option -- that the constructor code decides when to apply the init values. It is less code than the current typically used option, of simply setting the `init` value to `null` and setting the property's value in the constructor, particularly if there are many properties, but kind of papers over the very intended use of initializers. 

If I do end up calling the `apply` method and firing the change event automatically via one of these three options, for properties initialized with `initFunction`, I would provide arguments of `value, old` as `returnValueFromInitFunction, undefined`. By providing `old` as `undefined`, it allows the `apply` method or event listener to distinguish that this is an initial value, because there is no other time that `old` can be `undefined`.

Comments welcome!

**Comment by johnspackman** [comment]
## Legacy implementation

IMHO the `initXxx()` sounds pretty horrible, especially when you can change the init value.  Part of the purpose of a property declaration (whether the original intention or not) is to document what that property is and does, and it is pretty bad to clearly state "oh the init value is X" and then have the constructor or other code silently make it "Y", without an event or apply method - arguably this effectively allows future code (or derived code) to fundamentally change the implementation of a property, without changing the published API (or even breaking the published API).  

I've never understood the reason why sometimes you have to add `initXxx` in your constructor and most of the times not, I just knew "stuff happens" and been able to avoid digging into it.

I also didn't realise that the first call to `setXxxx` was special and would trigger apply/event even if the old and new values were the same - IMHO this does not make rational sense either.

An initial value sounds so simple (as it should be) but having to read a load of documentation to understand it properly is a big red flag that it's not a good design - my vote is that we take advantage of @derrell 's rewriting to clean this mess up and take the hit of some BC changes, which are presumably edge cases anyway.

## Future implementation

The problem that these init rules seem to be trying to address is that we do not want to have uninitialised properties *ever* (ie not even during constructors) yet at the same time the constructor may need to determine that initialisation.

These init rules will never solve all scenarios, but have the expense of being complicated and non-obvious.

I think that there is no need for special init procedures, because an init value should be exactly that - the property's initial value.  If you need to do something special, then you declare the initial value as `undefined` or `null` and then act appropriately in the constructor; or you can use the `initFunction` to create that special initial value that you need.  

At first, I thought that this might open up a gap where a property is uninitialised, but that's OK too - EG if a property is declared without an `init` value, it will have a value of `undefined` - and any time that `getXxx` is called an exception is raised instead of returning `undefined`.  This does mean that `initXxx` should still exist, but it should only be allowed to be called exactly once, and if the property has an `init` value then the `initXxx` function should either not exist or throw an exception when called.

I don't think that the return value from `initFunction` is special, it should not call apply/event any more than any other init value

This makes everything simple and obvious; even the `initXxxx` function disappears for most properties.

(Ideally I'd suggest replacing all the `initXxx` methods with `initPropertyValue("xxx", newValue);` but that's not necessary)

WRT to `resetXxx` that should reset it to its initial value - and if that means resetting it to `undefined` because there is no init value, then that is the correct course of action because we're following the developer's code and being obvious in our implementation.

PS - what is the use case for `refreshXxxx`?

**Reply by derrell** [reply]
The `initX` method with an argument is intended to be called from the constructor when the intention is that the `init` value is a reference type. It could theoretically be obsoleted by the new `initFunction` property configuration key... but it's likely widely used, so should probably be retained.

I see little (no) value in ever calling `initX` without an argument. I think it was supposed to allow specifying a reference type `init` value in the property configuration, but that doesn't work. The reference type is created as the property configuration is read, not when `initX` is executed.

There is a difference, in the qooxdoo v7 class/property system, between calling `prop.initX(23)` and calling `prop.setX(23)`. The former does not generate a call to the `apply` method or fire an event (at least not immediately; it may on the subsequent `setX` call), as does the latter; and the former also saves the internally-stored `init` value, allowing `prop.resetX` to set the value back to what was provided to `prop.initX`.

Reviewing the documentation, it seems that the argument to `initX` is supposed to be ignored if an `init` value is provided in the property configuration.

`refreshX()` is used by the property inheritance system.

**Reply by derrell** [reply]
No, experimentation and review of the generated code proves me wrong. Calling `initX()` with no arguments uses the `init` value, and causes the `apply` method to be called and the event to be fired, if the `init` value differs from the "old" value. The "old" value is, it seems, `undefined` if `deferredInit : true` was set in the property configuration; otherwise, it's the `init` value from the configuration.

There is an internal `$$initialized` flag after an instance's constructor is called. The `initX()` code throws an error if `initX()` is called when `$$initialized` is true, i.e., after the constructor has been run. This means that `initX()` may only be called from the constructor.

So the `initX` feature is what I said above that I was looking for: a way to cause `init` values (or in my description, `initFunction` return values) to somehow be used in calls to the `apply` method when initializing the class, which can allow a single place -- that `apply` method -- for processing what to do when a property value is set, even with its initial value.

**Comment by derrell** [comment]
In order to find out the magnitude of work required, just within the framework itself, if we decided to deprecate or eliminate the use of `initX()` method calls, I added a potential warning, using `console.warn`, to show all of the unique places where `initX()` is called. Here is the result of running the framework unit test suite:

```
calling initLocale is obsolete: construct@http://localhost:8080/transpiled/qx/locale/Manager.js:94:12
calling initCssClass is obsolete: construct@http://localhost:8080/transpiled/qx/html/Element.js:138:12
calling initFocusable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/core/Widget.js:121:12
calling initSelectable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/core/Widget.js:122:12
calling initNativeContextMenu is obsolete: construct@http://localhost:8080/transpiled/qx/ui/core/Widget.js:123:12
calling initNativeHelp is obsolete: construct@http://localhost:8080/transpiled/qx/ui/root/Abstract.js:97:12
calling initItemHeight is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/VirtualTree.js:205:12
calling initOpenMode is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/VirtualTree.js:206:12
calling initOpen is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/core/FolderOpenButton.js:51:12
calling initOpen is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/core/AbstractItem.js:80:12
calling initVisibility is obsolete: construct@http://localhost:8080/transpiled/qx/ui/popup/Popup.js:60:12
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/List.js:103:14
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/Slider.js:129:14
calling initValue is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/Spinner.js:153:14
calling initWrap is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/TextArea.js:81:12
calling initVisibility is obsolete: construct@http://localhost:8080/transpiled/qx/ui/menu/Menu.js:89:12
calling initKeepFocus is obsolete: construct@http://localhost:8080/transpiled/qx/ui/menu/Menu.js:90:12
calling initKeepActive is obsolete: construct@http://localhost:8080/transpiled/qx/ui/menu/Menu.js:91:12
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/container/SlideBar.js:113:14
calling initVisibility is obsolete: construct@http://localhost:8080/transpiled/qx/ui/window/Window.js:146:12
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/core/scroll/ScrollBar.js:99:14
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/core/scroll/NativeScrollBar.js:109:14
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/splitpane/Pane.js:73:14
calling initBarPosition is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tabview/TabView.js:100:14
calling initShowCloseButton is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tabview/TabButton.js:70:12
calling initOpenMode is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/Tree.js:105:12
calling initRootOpenClose is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/Tree.js:106:12
calling initEnabled is obsolete: construct@http://localhost:8080/transpiled/qx/bom/Shortcut.js:77:12
calling initGroups is obsolete: construct@http://localhost:8080/transpiled/qx/ui/list/List.js:110:12
calling initItemHeight is obsolete: construct@http://localhost:8080/transpiled/qx/ui/list/List.js:116:12
calling initSelection is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/core/VirtualDropDownList.js:79:12
calling initModel is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/core/AbstractVirtualBox.js:86:14
calling initSelection is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/VirtualSelectBox.js:91:12
calling initHtmlMarkers is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/VirtualSelectBox.js:102:12
calling initModel is obsolete: construct@http://localhost:8080/transpiled/qx/ui/form/core/AbstractVirtualBox.js:83:14
calling initModel is obsolete: construct@http://localhost:8080/transpiled/qx/ui/list/List.js:113:14
calling initShowCellFocusIndicator is obsolete: construct@http://localhost:8080/transpiled/qx/ui/table/pane/Scroller.js:211:12
calling initScrollTimeout is obsolete: construct@http://localhost:8080/transpiled/qx/ui/table/pane/Scroller.js:221:12
calling initStatusBarVisible is obsolete: construct@http://localhost:8080/transpiled/qx/ui/table/Table.js:255:12
calling initModel is obsolete: construct@http://localhost:8080/transpiled/qx/ui/tree/VirtualTree.js:202:14
calling initAppearance is obsolete: construct@http://localhost:8080/transpiled/qx/ui/virtual/cell/Cell.js:62:12
calling initInitApplyProp1 is obsolete: testInit/inst<@http://localhost:8080/transpiled/qx/test/core/Property.js:236:16
calling initInitApplyProp2 is obsolete: testInit/inst<@http://localhost:8080/transpiled/qx/test/core/Property.js:239:16
calling initPropTwo is obsolete: construct@http://localhost:8080/transpiled/qx/test/core/Property.js:802:18
calling initDefaultCssClass is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/core/Widget.js:97:12
calling initName is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/core/Widget.js:98:12
calling initAnonymous is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/core/Widget.js:99:12
calling initActivatable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/core/Widget.js:100:12
calling initWrap is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/basic/Label.js:84:12
calling initSelectable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/list/renderer/Abstract.js:48:12
calling initRemovable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/list/renderer/Abstract.js:49:12
calling initShowArrow is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/list/renderer/Abstract.js:50:12
calling initSource is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/basic/Image.js:78:14
calling initSelectable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/list/renderer/group/Abstract.js:43:12
calling initMaxLength is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/form/MText.js:66:12
calling initPlaceholder is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/form/MText.js:67:12
calling initReadOnly is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/form/MText.js:68:12
calling initSelectable is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/form/Row.js:52:12
calling initWrap is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/form/Label.js:119:12
calling initOrientation is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/container/Drawer.js:86:12
calling initPositionZ is obsolete: construct@http://localhost:8080/transpiled/qx/ui/mobile/container/Drawer.js:87:12
```

**Comment by johnspackman** [comment]
> No ... Calling initX() with no arguments uses the init value, and causes the apply method to be called and the event to be fired, if the init value differs from the "old" value. The "old" value is, it seems, undefined if deferredInit : true was set in the property configuration; otherwise, it's the init value from the configuration.

Whats the difference between `initX()` and `setX()`, apart from the `$$initialized` flag preventing the `initX()` method being called outside of the constructor?

**Reply by derrell** [reply]
This is my understanding:

If `initX()` is called with an argument, it sets the internal storage of the initial value to that argument, if there was no already stored initial value, i.e., from an `init : value` member in the property configuration. That internally stored value is then used as the new value in a call to `applyX()`, and an event fired. The internally stored value will also be used as the value to set the property to, whenever `resetX()` is called.

`setX()` uses its argument as the new value in a call to `applyX()` and an event fired, if the `isEqual()` method, passed the argument and the current value of the property, returns false. It does not affect the internally stored initial value, so the argument value will be lost whenever `resetX()` is called.

**Comment by johnspackman** [comment]
IMHO the `initX` function is unnecessary, confusing, and should be removed altogether; if a developer requires the same functionality, they can simply use `setX` instead, and in the short term have a temporary implementation of `initX` that defers to `setX` with a deprecated warning.  

The initial value (whether provided by the property definition, `deferredInit`, or `initFunction`) should *not* trigger apply/event.

**Comment by level420** [comment]
But what if someone wants to change "the value which is restored if `resetX` is called" to a particular instance of a class? 
I thought that's what `initX` is used for when called with an argument.

**Reply by johnspackman** [reply]
ah yes, you're right.

Looking at the framework code, it looks like `.initX()` (ie without any arguments) is being used (abused?) to trigger the apply/event for the initial value.  A quick grep suggests that there are a *lot* more examples of no arguments than with arguments.

It seems then that the decision on whether a class will be written to call `initX` depends on whether there is code in the `apply` method that is not reusable; in which case it would follow that a seemingly benign change to the implementation of the code effectively has a change on the class' public API, because it changes how the events are fired.

What I think is really wrong is to fire an event/apply method if `initX` is called, because no other form of initialisation will fire it.

**Reply by derrell** [reply]
I am getting closer to a completely BC implementation of what we've had traditionally. That said, I think there are changes that should be made _at some point_. Assuming I can get a completely BC implementation working right now, I propose the following as future changes, allowing time to properly deprecate things.

Premise:

- The addition of `initFunction` in the property configuration eliminates the need for `initX()` with an argument, for initializing a property with a reference type, which couldn't be done in the property configuration.
- The use of `initX()` without an argument, as @johnspackman points out, is there solely to cause the `apply` method to be called and event generated. I believe strongly that this capability should be available, as it allows a single place where actions occur -- the `apply` method -- that is used both for the initial value and for subsequent changes with `setX()`. I do believe that `initX()` is misnamed for that purpose, however.

Proposal:

- Deprecate now, but do not yet eliminate, `initX`. In the deprecation message, if possible, link to documentation discussing `initFunction` for initializing with reference types, and the below-described `applyInit` method.
- To allow for the ability to call the `apply` method for properties' initial values, I would add a new method:
```javascript
/**
 * Call the apply method and fire events for the specified properties,
 * using the return value from `initFunction` or the `init` value as the `value`
 * argument, and `undefined` as the `old` argument.
 *  
 * @param propertyNames {Array[String]}
 *   For each property name specified in this array, that property's
 *   `apply` method will be called, and its change event fired.
 *
 * @return {undefined}
 *   There is no provision for awaiting promise resolution nor to retrieve
 *   values back from the `apply` methods.
 */
applyInit(propertyNames)
{
}
```

The `applyInit` method would be available only during the time that the constructor is executing. Developers may choose to call it from their constructor, passing the list of properties that they would like the `apply` method called and change event fired. `applyInit` may be called multiple times, to allow for different properties to be `apply`ed at different places within the constructor, but the implementation of `applyInit` will throw an error if a property name is used in more than one call to `applyInit`.

**Reply by derrell** [reply]
@level420 According to the documentation, you can not use the `initX` method with an argument to change an init value:

```
If an init value is given in the property declaration, the init method does not accept any parameters.
The init methods always use the predefined init values. In cases where there is no init value given
in the property declaration, it is possible to call the init method with one parameter, which
represents the init value. This may be useful to apply reference types to each instance. Thus
they would not be shared between instances.
```
